lxd/metrics.go

```diff
@@ -0,0 +1,146 @@
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/lxd/metrics"
+	"github.com/lxc/lxd/lxd/response"
+	log "github.com/lxc/lxd/shared/log15"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+var metricsCmd = APIEndpoint{
+	Path: "metrics",
+
+	Get: APIEndpointAction{Handler: metricsGet, AccessHandler: allowProjectPermission("containers", "view")},
+}
+
+var metricsGetMu sync.Mutex
+var metricsGetBuilding bool
+
+// swagger:operation GET /1.0/metrics metrics metrics_get
+//
+// Get metrics
+//
+// Gets metrics of instances.
+//
+// ---
+// produces:
+//   - text/plain
+// parameters:
+//   - in: query
+//     name: project
+//     description: Project name
+//     type: string
+//     example: default
+// responses:
+//   "200":
+//     description: Metrics
+//     schema:
+//       type: string
+//       description: Instance metrics
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func metricsGet(d *Daemon, r *http.Request) response.Response {
+	// Prevent concurrent comparison/modification of metricsLastFetchedTime and access of d.metrics.
+	metricsGetMu.Lock()
+
+	// Return cached metrics build in progress or if they have been built in the last 15 seconds.
+	if d.metrics != nil && (metricsGetBuilding || d.metricsLastFetchedTime.Add(15*time.Second).After(time.Now())) {
+		metricsStr := d.metrics.String()
+		metricsGetMu.Unlock()
+
+		return response.SyncResponsePlain(true, metricsStr)
+	}
+
+	// Start metrics build and release lock so other requests can access cached metrics whilst we build.
+	// Access to d.metricsLastBuildTime and d.metrics is forbidden from here on until we lock again later,
+	// so the new metrics set will be built locally and stored in d.metrics at the end.
+	metricsGetBuilding = true
+	metricsGetMu.Unlock()
+
+	var (
+		dbInstances []db.Instance
+		err         error
+	)
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		filter := db.InstanceTypeFilter(instancetype.Any)
+
+		projectName := queryParam(r, "project")
+		if projectName != "" {
+			filter.Project = &projectName
+		}
+
+		dbInstances, err = tx.GetInstances(filter)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	instances, err := instance.LoadAllInternal(d.State(), dbInstances)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	metricsToMerge := make(chan *metrics.MetricSet)
+
+	// Launch concurrent gather of metrics from instances.
+	instCount := 0
+
+	for _, inst := range instances {
+		// Ignore stopped instances
+		if !inst.IsRunning() {
+			continue
+		}
+
+		instCount++
+
+		go func(inst instance.Instance) {
+			instanceMetrics, err := inst.Metrics()
+			if err != nil {
+				metricsToMerge <- nil
+				logger.Warn("Failed to get metrics", log.Ctx{"inst": inst.Name(), "err": err})
```

Comment:
@monstermunchkin need to add a project field here too, as the same instance name can exist in multiple projects.

Reply:
"Failed to get instance metrics" for clarity.